autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+        plotting individual runs
+        """
+        self.performance_tracker['train_loss'][epoch] = train_loss
+        self.performance_tracker['val_loss'][epoch] = val_loss
+        self.performance_tracker['test_loss'][epoch] = test_loss
+        self.performance_tracker['start_time'][epoch] = start_time
+        self.performance_tracker['end_time'][epoch] = end_time
+        self.performance_tracker['train_metrics'][epoch] = train_metrics
+        self.performance_tracker['val_metrics'][epoch] = val_metrics
+        self.performance_tracker['test_metrics'][epoch] = test_metrics
+
+    def register_model_statistics(self,
+                                  total_parameter_count: float,
+                                  trainable_parameter_count: float,
+                                  ) -> None:
+        """
+        Utility to register global statistic about the model or training,
+        that are agnostic to epochs
+        """
+        self.total_parameter_count = total_parameter_count
+        self.trainable_parameter_count = trainable_parameter_count
+
+    def get_best_epoch(self, loss_type: str = 'val_loss') -> int:
+        return np.argmin(
+            [self.performance_tracker[loss_type][e] for e in range(1, len(
+                self.performance_tracker[loss_type]) + 1
+            )]
+        )
+
+    def get_last_epoch(self) -> int:
+        if 'train_loss' not in self.performance_tracker:
+            return 0
+        else:
+            return max(self.performance_tracker['train_loss'].keys())
+
+    def repr_last_epoch(self) -> str:
+        """
+        For debug purposes, returns a nice representation of last epoch
+        performance
+
+        Returns:
+            str: A nice representation of the last epoch
+        """
+        last_epoch = len(self.performance_tracker['train_loss'])
+        string = "\n"
+        string += '=' * 40
+        string += f"\n\t\tEpoch {last_epoch}\n"
+        string += '=' * 40
+        string += "\n"
+        for key, value in sorted(self.performance_tracker.items()):
+            if isinstance(value[last_epoch], dict):
+                # Several metrics can be passed
+                string += "\t{}:\n".format(
+                    key,
+                )
+                for sub_key, sub_value in sorted(value[last_epoch].items()):
+                    string += "\t\t{}: {}\n".format(
+                        sub_key,
+                        sub_value,
+                    )
+            else:
+                string += "\t{}: {}\n".format(
+                    key,
+                    value[last_epoch],
+                )
+        string += '=' * 40
+        return string
+
+
+class BaseTrainerComponent(autoPyTorchTrainingComponent):
+
+    def prepare(
+        self,
+        metrics: List[Any],
+        model: torch.nn.Module,
+        criterion: torch.nn.Module,
+        budget_tracker: BudgetTracker,
+        optimizer: torch.nn.Module,
```

Comment:
I thought that torch.optim.optimizer wasnt an nn.Module (https://pytorch.org/docs/stable/_modules/torch/optim/optimizer.html#Optimizer) or am I wrong?

Reply:
Thanks for catching this, so much for static checking from mypy :D